File: test/cpu/tas.cc
#define BOOST_TEST_DYN_LINK
#include "68040.hpp"
#include "test.hpp"

#include <boost/test/data/monomorphic.hpp>
#include <boost/test/data/test_case.hpp>
#include <boost/test/unit_test.hpp>
namespace bdata = boost::unit_test::data;
struct F_TAS {
    F_TAS() {
        // TAS %D2
        TEST::SET_W(0, 0045300 | 2);
        TEST::SET_W(2, TEST_BREAK);
        jit_compile(0, 4);
    }};
BOOST_FIXTURE_TEST_SUITE(TAS, Prepare, *boost::unit_test::fixture<F_TAS>())


BOOST_AUTO_TEST_CASE(value) {
    cpu.D[2] = 1;
    run_test(0);
    BOOST_TEST(cpu.D[2] == 0x81);
    BOOST_TEST(!cpu.Z);
    BOOST_TEST(!cpu.V);
    BOOST_TEST(!cpu.N);
    BOOST_TEST(!cpu.C);
}

BOOST_AUTO_TEST_CASE(N) {
    cpu.D[2] = 0x81;
    run_test(0);
    BOOST_TEST(cpu.N);
}

BOOST_AUTO_TEST_CASE(Z) {
    cpu.D[2] = 0;
    run_test(0);
    BOOST_TEST(cpu.Z);
}

BOOST_AUTO_TEST_SUITE_END()

File: test/lua/cpu/tas.lua
-- TAS %D2
MEM[0].W = OCT "045300" | 2
MEM[2].W = TEST.BREAK

JIT_COMPILE(0, 4)

local run = {}

function run.value()
    CPU.D[3].B = 1
    TEST.RUN(0)
    TEST.CHECK("CPU.D[3].B", 0x81)
    TEST.CHECK("CPU.CC.V", false)
    TEST.CHECK("CPU.CC.C", false)
end

run.n = farray(TF, function(b)
    CPU.D[3].B = cond(b, 0x80, 0)
    TEST.RUN(0)
    TEST.CHECK("CPU.CC.N", b)
end)
run.z = farray(TF, function(b)
    CPU.D[3].B = cond(b, 0, 1)
    TEST.RUN(0)
    TEST.CHECK("CPU.CC.Z", b)
end)
return run
